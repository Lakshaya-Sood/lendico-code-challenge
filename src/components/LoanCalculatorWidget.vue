<template>
  <b-card header="Loan Calculator Widget" align="center">
    <b-card>
      <h4>Years</h4>
      <vue-slider
        :marks="true"
        :dotSize="20"
        @change="this.onYearChange"
        :min="this.yearRange.min"
        :max="this.yearRange.max"
        :value="this.yearsValue"
      ></vue-slider>
      <br />
    </b-card>
    <br />
    <b-card>
      <h4>Amount in €</h4>
      <b-form-input
        id="input-live"
        type="number"
        placeholder="Enter Amount"
        aria-describedby="input-live-help valid-feedback invalid-feedback"
        @change="this.onAmountChange"
        :state="isAmountValid"
        :value="this.amountValue"
      ></b-form-input>
      <b-form-invalid-feedback id="invalid-feedback">Amount not in specified Range</b-form-invalid-feedback>
      <b-form-valid-feedback id="valid-feedback">Amount in specified Range</b-form-valid-feedback>
      <b-form-text
        id="input-live-help"
      >Enter amount between €{{this.amountRange.min | formatPrice}} - €{{this.amountRange.max | formatPrice}}</b-form-text>
    </b-card>
    <br />
    <b-button pill size="lg" variant="dark" @click="this.handleSubmit">Calculate</b-button>
  </b-card>
</template>

<script lang='ts'>
import Vue from "vue";
import { Prop, Component } from "vue-property-decorator";
import { NumericRange } from "../typings/components";

const emptyfunc = () => {};
@Component({
  name: "loan-calculator-widget",
  filters: {
    formatPrice: function(price: number) {
      let formatted = new Intl.NumberFormat("de-DE", {
        currency: "EUR"
      }).format(price);
      return formatted;
    }
  }
})
export default class LoanCalculatorWidget extends Vue {
  @Prop({ required: true }) readonly yearRange!: NumericRange;
  @Prop({ required: true }) readonly amountRange!: NumericRange;
  @Prop({ required: true }) readonly yearsValue!: number;
  @Prop({ required: true }) readonly amountValue!: number;
  @Prop({ default: emptyfunc }) readonly onYearChange!: (value: number) => void;
  @Prop({ default: emptyfunc }) readonly onAmountChange!: (
    value: number
  ) => void;
  @Prop({ default: emptyfunc }) readonly onSubmit!: () => void;

  //computed
  get isAmountValid() {
    let { min, max } = this.amountRange;
    return this.amountValue >= min && this.amountValue <= max;
  }

  //methods
  handleSubmit() {
    this.isAmountValid
      ? this.$emit("onSubmit")
      : this.$bvToast.toast("Please fill the correct values", {
          title: `Form Submit`,
          variant: "danger",
          autoHideDelay: 1500,
          solid: true
        });
  }
}
</script>

