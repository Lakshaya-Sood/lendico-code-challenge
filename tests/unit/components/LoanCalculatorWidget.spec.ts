import { mount, createLocalVue } from '@vue/test-utils'
import BootstrapVue from 'bootstrap-vue'
import VueSlider from "vue-slider-component";

import LoanCalculatorWidget from '@/components/LoanCalculatorWidget.vue'

const vue_mock = createLocalVue()
vue_mock.use(BootstrapVue)
vue_mock.component('VueSlider', VueSlider)


describe('loan-calculator-widget  component', () => {
    const mocks = {
        yearsValue: 2,
        yearRange: { min: 1, max: 5 },
        amountRange: { min: 10, max: 100 },
        onYearChange: (value: number) => { },
        onAmountChange: (value: number) => { },
        onSubmit: jest.fn()
    }
    const wrapper_valid_feilds = mount(LoanCalculatorWidget, {
        propsData: {
            amountValue: 80,
            ...mocks
        },
        localVue: vue_mock
    });
    const wrapper_invalid_feilds = mount(LoanCalculatorWidget, {
        propsData: {
            amountValue: 120,
            ...mocks
        },
        localVue: vue_mock
    });
    const spyOnToast = jest.spyOn(wrapper_invalid_feilds.vm.$bvToast, 'toast')

    test('name in kebab case', () => {
        expect(wrapper_valid_feilds.vm.$options.name).toEqual('loan-calculator-widget')
    })

    describe('[method] handleSubmit()', () => {
        test('valid feilds', () => {
            wrapper_valid_feilds.find('button').trigger('click')
            expect(wrapper_valid_feilds.emitted().onSubmit).toBeTruthy()
        })
        test('invalid feilds', () => {
            wrapper_invalid_feilds.find('button').trigger('click')
            expect(wrapper_invalid_feilds.emitted().onSubmit).toBeFalsy()
            expect(spyOnToast).toHaveBeenCalled();
        })
    })
    test('no changes as compared to last snapshot', () => {
        expect(wrapper_valid_feilds.element).toMatchSnapshot() // https://vue-test-utils.vuejs.org/guides/#snapshot-testing
    })
})
