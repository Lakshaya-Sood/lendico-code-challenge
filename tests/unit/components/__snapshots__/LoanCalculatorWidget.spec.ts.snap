// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loan-calculator-widget  component no changes as compared to last snapshot 1`] = `
<div
  class="card text-center"
>
  <!---->
  <div
    class="card-header"
  >
    <div>
      Loan Calculator Widget
    </div>
  </div>
  <div
    class="card-body"
  >
    <!---->
    <!---->
    <div
      class="card"
    >
      <!---->
      <!---->
      <div
        class="card-body"
      >
        <!---->
        <!---->
        <h4>
          Years
        </h4>
         
        <div
          class="vue-slider vue-slider-ltr"
          style="padding: 10px 0px; width: auto; height: 4px;"
        >
          <div
            class="vue-slider-rail"
          >
            <div
              class="vue-slider-process"
              style="height: 100%; top: 0px; left: 0%; width: 25%; transition-property: width,left; transition-duration: 0.5s;"
            />
            <div
              class="vue-slider-marks"
            >
              <div
                class="vue-slider-mark vue-slider-mark-active"
                style="height: 100%; width: 4px; left: 0%;"
              >
                <div
                  class="vue-slider-mark-step vue-slider-mark-step-active"
                />
                <div
                  class="vue-slider-mark-label vue-slider-mark-label-active"
                >
                  1
                </div>
              </div>
              <div
                class="vue-slider-mark vue-slider-mark-active"
                style="height: 100%; width: 4px; left: 25%;"
              >
                <div
                  class="vue-slider-mark-step vue-slider-mark-step-active"
                />
                <div
                  class="vue-slider-mark-label vue-slider-mark-label-active"
                >
                  2
                </div>
              </div>
              <div
                class="vue-slider-mark"
                style="height: 100%; width: 4px; left: 50%;"
              >
                <div
                  class="vue-slider-mark-step"
                />
                <div
                  class="vue-slider-mark-label"
                >
                  3
                </div>
              </div>
              <div
                class="vue-slider-mark"
                style="height: 100%; width: 4px; left: 75%;"
              >
                <div
                  class="vue-slider-mark-step"
                />
                <div
                  class="vue-slider-mark-label"
                >
                  4
                </div>
              </div>
              <div
                class="vue-slider-mark"
                style="height: 100%; width: 4px; left: 100%;"
              >
                <div
                  class="vue-slider-mark-step"
                />
                <div
                  class="vue-slider-mark-label"
                >
                  5
                </div>
              </div>
            </div>
            <div
              class="vue-slider-dot"
              style="width: 20px; height: 20px; transform: translate(-50%, -50%); top: 50%; left: 25%; transition: left 0.5s;"
            >
              <div
                class="vue-slider-dot-handle"
              />
              <div
                class="vue-slider-dot-tooltip vue-slider-dot-tooltip-top"
              >
                <div
                  class="vue-slider-dot-tooltip-inner vue-slider-dot-tooltip-inner-top"
                >
                  <span
                    class="vue-slider-dot-tooltip-text"
                  >
                    2
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
         
        <br />
      </div>
      <!---->
      <!---->
    </div>
     
    <br />
     
    <div
      class="card"
    >
      <!---->
      <!---->
      <div
        class="card-body"
      >
        <!---->
        <!---->
        <h4>
          Amount in €
        </h4>
         
        <input
          aria-describedby="input-live-help valid-feedback invalid-feedback"
          class="form-control is-valid"
          id="input-live"
          placeholder="Enter Amount"
          type="number"
        />
         
        <div
          class="invalid-feedback"
          id="invalid-feedback"
        >
          Amount not in specified Range
        </div>
         
        <div
          class="valid-feedback"
          id="valid-feedback"
        >
          Amount in specified Range
        </div>
         
        <small
          class="form-text text-muted"
          id="input-live-help"
        >
          Enter amount between €10 - €100
        </small>
      </div>
      <!---->
      <!---->
    </div>
     
    <br />
     
    <button
      class="btn btn-dark btn-lg rounded-pill"
      type="button"
    >
      Calculate
    </button>
  </div>
  <!---->
  <!---->
</div>
`;
